Sources/MessagingPush/PushHandling/ManualPushHandling+UserNotifications.swift

```diff
@@ -96,7 +96,9 @@ extension MessagingPushImplementation {
         // A hack to get an instance of pushClickHandler without making it a property of the MessagingPushImplementation class. pushClickHandler is not available to app extensions but MessagingPushImplementation is.
         // We get around this by getting a instance in this function, only.
         if let pushClickHandler = sdkInitializedUtil.postInitializedData?.diGraph.pushClickHandler {
-            pushClickHandler.pushClicked(push)
+            pushClickHandler.trackPushMetrics(for: push)
+            pushClickHandler.cleanupAfterPushInteractedWith(for: push)
+            pushClickHandler.handleDeepLink(for: push)
```

Comment:
does the order matter? i would think so? as `handleDeepLink` might redirect to another app. 

what are we doing to avoid a case where someone might call one method before another? specially when the method is named `cleanup` which would confuse anyone to call at the end?

Reply:
if the above supposition is true, then it would incline towards the previous [suggestion](https://github.com/customerio/customerio-ios/pull/590#discussion_r1516694859) 

We should have another method that dictates the ordering of methods, which automated test might not be able to do

Reply:
Great call-out. The order is important. 

From reading code in the link you sent, I may not fully understand the suggestion, but I do not think that the solution will solve the problem we have with multiple parts of the code performing click tracking and the functions being called are unique. 

I think that automated tests are the best place to assert the ordering. I will add this functionality to the tests.

Reply:
[I updated the test suite to assert order](https://github.com/customerio/customerio-ios/compare/3f3fa43cca68798ae4baf726c27d654dd1fc43dd..660a9906d8a8e1a2bc88d1867a4922a43d31c642). 

Notice that I added a `XXX:` suggestion to the test function. To keep the scope small, I did not implement this. If this sounds like a good idea to do, I can create a future PR to do this.

Reply:
Great test, I think this test would work for now but my concern regarding the usage of this class remains. This test is great for the methods we already know are utilizing push click handling. 

My concern is, that someone adds another method in `PushEventHandler`, in that method if they are supposed to do push handling, and don't follow the order, we can't stop it. But if we had a method that they can call, that respects the order for them, it would have made it simpler.

Reply:
> But if we had a method that they can call, that respects the order for them, it would have made it simpler.

Previously, we did have 1 function that could be called and all push click handling could be done inside of it: `PushClickHandler.onPushClicked`. 

This PR made this more complex because now the push click handling when we do automatic and manual push click handling is different. The order of operations is identical, but the implementations are still unique. Therefore, I think we would need to implement 2 versions of `onPushClicked`, right? 

Assuming this is what you're trying to say, I think that leveraging automated tests to verify push click handling is what we expect across the whole SDK, is a better solution then using multiple `PushClickHandler` functions that perform all of the push click handling. 

Maybe I don't understand your comment?

Reply:
I am okay with moving forward with automated test for now, since its an internal API. I'll create proposal ticket as that might be out of scope for this.

Reply:
That sounds like a plan. 

I believe this is the last unresolved suggestion. Mind approving the PR to unblock merging it?

Reply:
```
    func assertWillHandleDeepLinkLast(for push: PushNotification) {
        // When deep link is handled, we expect all other click handling to have already been done.
        handleDeepLinkClosure = { _ in
            // We expect all other push click handling to have already happened.
            self.assertHandledPushClick(for: push)

            // XXX: This would be a good place to call: mock.verifyNoMoreInteractions(), like in Mockito, to be more confident that we are calling deep link last.
        }
    }
```

are you planing on adding the `XXX` part later? because as i understand this test doesnt currently verify that its called at last?

Reply:
> are you planing on adding the XXX part later?

To keep the scope small, I did not implement it in this PR. I was first going to look for feedback from the team in the PR to see if anyone had objections to it before I spend the effort on it. If no objections, I do plan on implementing this, yes. 

> as i understand this test doesnt currently verify that its called at last?

The test could be improved, but I do think that it verifies that it's called last. The function `assertWillHandleDeepLinkLast`, when deep link is handled, will check if all other push click handling was done. Which, is a form of checking that handling the deep link was done last. 

The test does not verify that no more push click handling operations are done after the deep link is handled. The `XXX:` comment calls this out and offers a potential solution.